lib/chef/provider/package/rubygems.rb

```diff
@@ -482,7 +482,7 @@ def cleanup_after_converge
 
         def candidate_version
           @candidate_version ||= begin
-                                   if target_version_already_installed?(@current_resource.version, @new_resource.version)
+                                   if action != :upgrade && target_version_already_installed?(@current_resource.version, @new_resource.version)
```

Comment:
Dropping the conditional would cause the spec [`"when determining the candidate version to install" it "does not query for available versions when the current version is the target version"`](https://github.com/wjordan/chef/blob/gem_upgrade_fix_3867/spec/unit/provider/package/rubygems_spec.rb#L494-L501) to fail. The fact that this behavior is explicitly tested makes me believe this implementation was intended, at least for the `:install` case (probably to avoid an extra RubyGems API round-trip when not necessary).